File: Data Structures & Algorithms/buy-and-sell-crypto/submission-1.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int l = 0;
        int r = 1;
        int best = 0;

        while (r < prices.size())
        {
            if (prices[l] < prices[r])
            {
                if (prices[r] - prices[l] > best)
                {
                    best = prices[r] - prices[l];
                }
            }
            else
            {
                l = r;
            }
            r++;
        }

        return best;
    }
};

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-1.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        memo = {}

        def recur(i, prev) -> int:
            if (i, prev) in memo:
                return memo[(i, prev)]
            
            best = 0
            
            for j in range(i, len(nums)):
                if nums[j] > prev:
                    best = max(best, 1 + recur(j, nums[j]))

            memo[(i, prev)] = best
            return best

        return recur(0, float('-inf'))